pkg/custom-plugins/plugin_output.go

```diff
@@ -0,0 +1,189 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert value for path %q to string: %w", jsonPath.QueryPath, err)
+		}
+
+		match, rule, err := jsonPath.MatchRule.checkMatchRule(strVal)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check match rule for path %q: %w", jsonPath.QueryPath, err)
+		}
+		results[jsonPath.FieldName] = extractedExtraInfo{
+			fieldName:  jsonPath.FieldName,
+			fieldValue: strVal,
+			matched:    match,
+			rule:       rule,
+		}
+	}
+
+	return results, nil
+}
+
+// anyToString converts any value to a string representation.
+// If the value is a complex type (map, slice), it's marshaled to JSON.
+func anyToString(v any) (string, error) {
+	if v == nil {
+		return "null", nil
+	}
+
+	switch val := v.(type) {
+	case string:
+		return val, nil
+	case bool, int, int64, float64, float32:
+		return fmt.Sprintf("%v", val), nil
+	default:
+		// For complex types (maps, slices, structs), marshal to JSON
+		jsonBytes, err := json.Marshal(val)
+		if err != nil {
+			return "", err
+		}
+		return string(jsonBytes), nil
+	}
+}
+
+// parseFirstJSON finds the start of the first JSON object in the input,
+// and parses it into a map[string]interface{}.
+// And if there are multiple JSON objects in the input, it only parses the first one.
+// The second and later ones are discarded.
+// If there is no JSON object, it returns nil and no error.
+func parseFirstJSON(input []byte) (map[string]any, error) {
+	// find the index of the first opening brace, which indicates the start of a JSON object
+	startIdx := bytes.IndexByte(input, '{')
+	if startIdx == -1 {
+		return nil, nil
+	}
+
+	// extract data starting from the first JSON object
+	jsonData := input[startIdx:]
+
+	// create a decoder to properly handle the JSON data
+	decoder := json.NewDecoder(bytes.NewReader(jsonData))
+	var rm map[string]any
+	if err := decoder.Decode(&rm); err != nil {
+		return nil, err
+	}
+
+	return rm, nil
+}
+
+// readJSONPath queries the target JSON path and returns the corresponding value,
+// in the input map. The returned value can be any type: string, nested map, float,
+// null, etc..
+// And most importantly, if the key is not found, it returns nil and no error.
+//
+// ref. https://pkg.go.dev/github.com/PaesslerAG/jsonpath#section-readme
+// ref. https://en.wikipedia.org/wiki/JSONPath
+// ref. https://goessner.net/articles/JsonPath/
+func readJSONPath(input map[string]any, path string) (any, error) {
+	v, err := jsonpath.Get(path, input)
+	if err != nil {
+		if strings.Contains(err.Error(), "unknown key") {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return v, nil
+}
+
+// checkMatchRule checks if the input matches the match rule.
+// It returns true if the input matches the match rule, otherwise false.
+// It returns an error if the match rule is invalid.
+// The second return value is the match rule that was used to check the input.
+func (rule MatchRule) checkMatchRule(input string) (bool, string, error) {
```

Comment:
Please put the method close to where MatchRule is defined at. That can improve the readability.  BTW  for the receiver type, we should use pointer by default.